crates/but/src/legacy/id/mod.rs

```diff
@@ -18,56 +17,95 @@ struct ContextInfo {
+    let (branch_names, commit_ids, worktree_dir, committed_files) = {
+        let repo = &*ctx.repo.get()?;
+        let head_info = but_workspace::head_info(
+            repo,
+            &meta,
+            but_workspace::ref_info::Options {
+                expensive_commit_info: false,
+                ..Default::default()
+            },
+        )?;
+        let mut branch_names: Vec<BString> = Vec::new();
+        let mut commit_ids: Vec<gix::ObjectId> = Vec::new();
+        let mut committed_files: Vec<(gix::ObjectId, BString)> = Vec::new();
+        for stack in head_info.stacks {
+            for segment in stack.segments {
+                if let Some(ref_info) = segment.ref_info {
+                    branch_names.push(ref_info.ref_name.shorten().to_owned());
+                }
+                for commit in segment.commits {
+                    let inner = commit.inner;
+                    commit_ids.push(inner.id);
+
+                    let tree_changes = but_core::diff::tree_changes(
+                        repo,
+                        inner.parent_ids.first().copied(),
+                        inner.id,
+                    )?;
+                    for tree_change in tree_changes {
+                        committed_files.push((inner.id, tree_change.path));
+                    }
+                }
+                for commit in segment.commits_on_remote {
+                    commit_ids.push(commit.id);
                 }
             }
-            for commit in segment.commits_on_remote {
-                commit_ids.push(commit.id);
-            }
+        }
+        committed_files.sort();
+
+        (
+            branch_names,
+            commit_ids,
+            repo.worktree().map(|worktree| worktree.base().to_owned()),
+            committed_files,
+        )
+    };
+    let mut uncommitted_files: Vec<(Option<StackId>, BString)> = Vec::new();
+    if let Some(worktree_dir) = worktree_dir {
+        let changes = but_core::diff::ui::worktree_changes_by_worktree_dir(worktree_dir)?.changes;
+        let (assignments, _assignments_error) =
+            but_hunk_assignment::assignments_with_fallback(ctx, false, Some(changes), None)?;
+        for assignment in assignments {
+            uncommitted_files.push((assignment.stack_id, assignment.path_bytes.clone()));
         }
     }
-    committed_files.sort();
+
     Ok(ContextInfo {
         branch_names,
         commit_ids,
+        uncommitted_files,
         committed_files,
     })
 }
 
+/// a.cmp(b) == a.id.cmp(&b.id) for all a and b
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+struct UncommittedFile {
+    assignment_path: (Option<StackId>, BString),
+    id: String,
+}
+impl Borrow<(Option<StackId>, BString)> for UncommittedFile {
+    fn borrow(&self) -> &(Option<StackId>, BString) {
+        &self.assignment_path
+    }
+}
+impl Borrow<str> for UncommittedFile {
+    fn borrow(&self) -> &str {
+        &self.id
+    }
+}
```

Comment:
I really like these `Borrow` implementations by the way and I wasn't aware of how this can be used to use sets a little like maps!
It's great.

CC @krlvi